### PR TITLE
Lint staged JS on pre-commit and run tests on pre-push

### PR DESCRIPTION
**Pull request by elektronik2k5**
As mentioned in https://github.com/ctrlplusb/react-universally/issues/358, this PR enforces:

1. `eslint` to run as a pre-commit hook on all staged js/jsx files.
2. `jest` to run as pre-push hook.

Should I also add documentation? Should it be an opt-in/out feature?

### CODE REVIEW

**Comment by sergiokopplin** [comment]
Is Yarn valid to add on num scripts? Some people don't use Yarn in development/production, just npm.

**Comment by ctrlplusb** [comment]
This is awesome, thanks @elektronik2k5. 🙏

I agree with @sergiokopplin, let's change those commands over to npm to save those who don't use yarn.

Really dig those libraries you are using. Starred! 😀

**Comment by elektronik2k5** [comment]
@sergiokopplin @ctrlplusb I discovered [yarn-or-npm](https://github.com/camacho/yarn-or-npm) and it seems to be working well with `yarn`. Didn't test it myself without yarn, but I assume it'll work.

**Comment by birkir** [comment]
I'm not entirely sure why not just use `npm` in stead of a third party package.

 - Most if not all build systems that support yarn, symlink npm to yarn if it is the target package manager (travis, heroku, google app engine, etc.).
 - If you run any scripts with yarn (eg. `yarn run lint`), it will automatically replace `npm` with yarn in subscripts (eg. `"lint": "npm run eslint; npm run stylelint"` will be executed like `yarn run eslint; yarn run stylelint`)

So using `npm` in scripts is the best solution.

**Comment by strues** [comment]
You're able to leave off the run portion of the command with yarn. For example `yarn lint`

**Comment by elektronik2k5** [comment]
First, I apologize for the upcoming wall of text!

@birkir,
> Most if not all build systems that support yarn, symlink npm to yarn if it is the target package manager (travis, heroku, google app engine, etc.).

Do you have any references to that?
Cause it sounds very wrong and broken: while `npm` and `yarn` share the same goal and many features, they are different and `yarn` is not (nor does it try to be) a drop-in replacement for `npm`.
They even have different CLI arguments for the same tasks! [They are different](https://yarnpkg.com/lang/en/docs/migrating-from-npm/#toc-cli-commands-comparison).

The only way I can imagine which would enable aliasing `yarn` to `npm` is by locking both their versions and maintaining translation aliases or shell functions from `npm` to `yarn`. But for it to work in a predictable and robust manner, one would have to actually parse all the arguments, which is a terrible idea. Am I missing something here?

In any case this isn't relevant to the current repo, since we want our scripts to **Just Work©**. Which brings me to your second point:
> If you run any scripts with yarn (eg. `yarn run lint`), it will automatically replace `npm` with yarn in subscripts (eg. `"lint": "npm run eslint; npm run stylelint"` will be executed like `yarn run eslint; yarn run stylelint`)

Unless there's some system wide aliasing, this does not work. Given this `package.json`:
```json
"scripts": {
  "who": "echo $(ps -o args= $PPID)",
  "who-npm": "npm run who",
  "who-yarn": "yarn run who"
}
```
Let's use the `"who"` script to print the parent process name to see what's executed.
### **`npm run who` output:**
```bash
> react-universally@13.0.0 who /home/elektronik/code/react-universally
> echo $(ps -o args= $PPID)
npm
```
_Note the last line_ - `npm`
### **`yarn run who` output:**
```bash
yarn run v0.20.3
$ echo $(ps -o args= $PPID) 
node /home/elektronik/.nodenv/versions/7.6.0/bin/yarn run who
Done in 0.24s.
```
_Note the last two lines_ - `yarn`

Now let's mix the two! :grin: 
### **`npm run who-npm` output:**
```bash
> react-universally@13.0.0 who-npm /home/elektronik/code/react-universally
> npm run who
> react-universally@13.0.0 who /home/elektronik/code/react-universally
> echo $(ps -o args= $PPID)
npm
```
### **`yarn run who-npm` output:**
```bash
yarn run v0.20.3
$ npm run who 
> react-universally@13.0.0 who /home/elektronik/code/react-universally
> echo $(ps -o args= $PPID)
npm
Done in 0.57s.
```
### **`npm run who-yarn` output:**
```bash
> react-universally@13.0.0 who-yarn /home/elektronik/code/react-universally
> yarn run who
yarn run v0.20.3
$ echo $(ps -o args= $PPID) 
node /home/elektronik/.nodenv/versions/7.6.0/bin/yarn run who
Done in 0.20s.
```
### **`yarn run who-yarn` output:**
```bash
yarn run v0.20.3
$ yarn run who 
yarn run v0.20.3
$ echo $(ps -o args= $PPID) 
node /home/elektronik/.nodenv/versions/7.6.0/bin/yarn run who
Done in 0.19s.
Done in 0.53s.
```

As you can see neither `npm` or `yarn` do anything but run shell commands. So if we want to support both, we need to either do what `yarn-or-npm` does or duplicate it's functionality. :disappointed:

**Comment by birkir** [comment]
Hey thanks for the detailed answer, my hats off to you.

I should have stated **I assumed©**, because clearly I'm wrong based on your findings. I based my answer on a personal experience with running scripts that didn't work with npm but worked with yarn, they did also work when yarn ran a npm subscript but not the other way around.

So fair enough. In this particular case, wouldn't using just npm be sufficient or will we get a different behaviour from yarn/npm?

**Comment by elektronik2k5** [comment]
@birkir I think there's value in having a single task runner. I consider even a minor thing such as output consistency (not mixing npm with yarn different output styles, like in some of the examples above) an important aspect of DX.

More practically, as a user who chooses one tool over another (regardless of which), I'd like to deal only with my choice (learning the features/api, debugging, updating it, etc.).

Basically, if I've moved to yarn, I don't ever want to worry or deal with npm :grimacing: .

**Comment by ctrlplusb** [comment]
Hi @elektronik2k5 

I am ready to get this merged in. Finally. :)

Do you mind sorting out the conflicts first?

Ta

**Comment by ctrlplusb** [comment]
Hey all!

So after all this I have decided to integrate `prettier` and `prettier-eslint`.  It auto-runs after each commit, so no need to do any precommit lint checking. 👍

I've added a prepush hook to do testing.

Thanks everyone involved here!
